appsec/src/helper/service.hpp

```diff
@@ -21,13 +24,109 @@ namespace dds {
 using namespace std::chrono_literals;
 
 class service {
-public:
+protected:
+    class MetricsImpl : public metrics::TelemetrySubmitter {
+        struct tel_metric {
+            tel_metric(std::string_view name, double value, std::string tags)
+                : name{name}, value{value}, tags{std::move(tags)}
+            {}
+            std::string_view name;
+            double value;
+            std::string tags;
+        };
+
+    public:
+        MetricsImpl() = default;
+        MetricsImpl(const MetricsImpl &) = delete;
+        MetricsImpl &operator=(const MetricsImpl &) = delete;
+        MetricsImpl(MetricsImpl &&) = delete;
+        MetricsImpl &operator=(MetricsImpl &&) = delete;
+
+        ~MetricsImpl() override = default;
+
+        void submit_metric(std::string_view metric_name, double value,
+            std::string tags) override
+        {
+            SPDLOG_TRACE("submit_metric: {} {} {}", metric_name, value, tags);
+            const std::lock_guard<std::mutex> lock{pending_metrics_mutex_};
+            pending_metrics_.emplace_back(metric_name, value, std::move(tags));
+        }
+
+        void submit_legacy_metric(std::string_view name, double value) override
+        {
+            SPDLOG_TRACE("submit_legacy_metric: {} {}", name, value);
+            const std::lock_guard<std::mutex> lock{legacy_metrics_mutex_};
+            legacy_metrics_[name] = value;
+        }
+        void submit_legacy_meta(
+            std::string_view name, std::string value) override
+        {
+            SPDLOG_TRACE("submit_legacy_meta: {} {}", name, value);
+            const std::lock_guard<std::mutex> lock{meta_mutex_};
+            meta_[std::string{name}] = std::move(value);
+        }
+
+        // NOLINTNEXTLINE(bugprone-easily-swappable-parameters)
+        void submit_legacy_meta_copy_key(
+            std::string name, std::string value) override
+        {
+            const std::lock_guard<std::mutex> lock{meta_mutex_};
+            meta_[std::move(name)] = std::move(value);
+        }
```

Comment:
I don't think you need two different names for these methods as if copy-semantics are used both methods are equivalent and move-semantics can only result in one of the two being called. On the other hand, shouldn't both log?

Reply:
That's technically true, but I'm making a distinction where std::string_view is used for static storage constants, so this helps reinforce that, even if the implementations don't then use separate maps for owned and non-owned keys (they could though).